lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1727,18 +1843,30 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       return conditionalStmt->rebindsSelf(Ctx, /*requireLoadExpr*/ true);
```

Comment:
> It's not quite related to this PR, but I'm not sure I understand the `requireLoadExpr` parameter of rebindsSelf. 

Thanks for noticing this, you're right that this check was previously over simplistic and unexpectedly allowed this example to compile:

```swift
var `self` = self ?? .somethingElse
guard let self = self else { return }
method() // this will be allowed
```

I updated the check to properly reject this sort of case. This change is source-breaking, so it emits a warning in Swift 5 mode and an error in Swift 6 mode.